### PR TITLE
[4.1] Fix Image Size Check with scheduler

### PR DESCRIPTION
**Pull request by bembelimen**
### Summary of Changes

- Fix the resizing of the image with the scheduler task
- Improve the description to make aware, that the file will be overwritten

### Testing Instructions

- Create a task from the type "Image Size Check"
- Run it

### Actual result BEFORE applying this Pull Request

Image sizes are not changed.

### Expected result AFTER applying this Pull Request
 Image sizes are changed.

### CODE REVIEW

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 4b49ad2ee00d579df2695075aebafb2eb69c3853<br /><br />Resizes the first image found that is too big but then says its completed<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36653">issues.joomla.org/tracker/joomla-cms/36653</a>.</sub>

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on 4b49ad2ee00d579df2695075aebafb2eb69c3853<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36653">issues.joomla.org/tracker/joomla-cms/36653</a>.</sub>

**Comment by bembelimen** [comment]
> I have tested this item red_circle unsuccessfully on [4b49ad2](https://github.com/joomla/joomla-cms/commit/4b49ad2ee00d579df2695075aebafb2eb69c3853)Resizes the first image found that is too big but then says its completed

It was always the intention to change one image per run to prevent any timeouts.

**Comment by brianteeman** [comment]
Then it must be clearly documented otherwise it will appear to be broken.

Although I have to say if it only does one image per run then it is pretty useless

File: administrator/language/en-GB/plg_task_checkfiles.ini [hunk 1]
```diff
@@ -11,6 +11,6 @@ PLG_TASK_CHECK_FILES_LOG_IMAGE_PATH_NA="Image path does exist!"
 PLG_TASK_CHECK_FILES_LOG_IMAGE_SAVE_FAIL="Failed to save image file"
 PLG_TASK_CHECK_FILES_LOG_RESIZE_FAIL="Failed to resize image due to an error in plugin logic..."
 PLG_TASK_CHECK_FILES_LOG_RESIZING_IMAGE="Found image of size %1$sx%2$s px; resizing to %3$sx%4$s px. File: %5$s"
-PLG_TASK_CHECK_FILES_TASK_IMAGE_SIZE_DESC="Check images, resize if larger than allowed."
+PLG_TASK_CHECK_FILES_TASK_IMAGE_SIZE_DESC="Check images, resize if larger than allowed. Attention: The original file will be overwritten!"
```
Comment (brianteeman) [inline administrator/language/en-GB/plg_task_checkfiles.ini:14]:
```suggestion
PLG_TASK_CHECK_FILES_TASK_IMAGE_SIZE_DESC="The selected folder will be checked for oversized images and the first image found will be resized. On each run of this task only one image will be resized. Attention: The original file will be overwritten!"
```

**Comment by bembelimen** [comment]
Updated the PR to allow more than one image per run.

File: administrator/language/en-GB/plg_task_checkfiles.ini [hunk 2]
```diff
@@ -5,12 +5,13 @@
 
 PLG_TASK_CHECK_FILES="Task - Check Files"
 PLG_TASK_CHECK_FILES_LABEL_DIMENSION_LIMIT="Limit (px)"
+PLG_TASK_CHECK_FILES_LABEL_DIMENSION_NUMIMAGES="Numer of Images"
```
Comment (brianteeman) [inline administrator/language/en-GB/plg_task_checkfiles.ini]:
```suggestion
PLG_TASK_CHECK_FILES_LABEL_DIMENSION_NUMIMAGES="Number of Images"
```

File: plugins/task/checkfiles/forms/image_size.xml [hunk 3]
```diff
@@ -29,10 +30,20 @@
...
-				min="100"
+				min="1"
 				step="1"
 				filter="int"
 			/>
+			<field
+				name="numImages"
+				type="number"
+				label="PLG_TASK_CHECK_FILES_LABEL_DIMENSION_NUMIMAGES"
+				required="true"
```
Comment (brianteeman) [inline plugins/task/checkfiles/forms/image_size.xml:42]: how about adding a description. something like

The maximum number of images to be processed in a single run. Setting this number too high may cause the task to timeout

File: administrator/language/en-GB/plg_task_checkfiles.ini [hunk 4]
```diff
@@ -5,12 +5,13 @@
...
 PLG_TASK_CHECK_FILES_LABEL_IMAGE_DIMENSION="Dimension"
 PLG_TASK_CHECK_FILES_LOG_IMAGE_PATH_NA="Image path does exist!"
 PLG_TASK_CHECK_FILES_LOG_IMAGE_SAVE_FAIL="Failed to save image file"
 PLG_TASK_CHECK_FILES_LOG_RESIZE_FAIL="Failed to resize image due to an error in plugin logic..."
 PLG_TASK_CHECK_FILES_LOG_RESIZING_IMAGE="Found image of size %1$sx%2$s px; resizing to %3$sx%4$s px. File: %5$s"
-PLG_TASK_CHECK_FILES_TASK_IMAGE_SIZE_DESC="Check images, resize if larger than allowed."
+PLG_TASK_CHECK_FILES_TASK_IMAGE_SIZE_DESC="Check images, resize if larger than allowed. Attention: The original file will be overwritten!"
 PLG_TASK_CHECK_FILES_TASK_IMAGE_SIZE_TITLE="Image Size Check"
-PLG_TASK_CHECK_FILES_XML_DESCRIPTION="Offers task routines for checking for oversized files, and related actions if possible."
+PLG_TASK_CHECK_FILES_XML_DESCRIPTION="Offers task routines for checking for oversized files, and related actions if possible."
```
Comment (tecpromotion) [inline administrator/language/en-GB/plg_task_checkfiles.ini]:
```suggestion
PLG_TASK_CHECK_FILES_XML_DESCRIPTION="Offers task routines for checking for oversized files, and related actions if possible."

```

Comment (tecpromotion) [reply administrator/language/en-GB/plg_task_checkfiles.ini]: add new line EOF

File: plugins/task/checkfiles/forms/image_size.xml (hunk 3 above)
Comment (bembelimen) [reply plugins/task/checkfiles/forms/image_size.xml:42]: Thx, added it.

File: administrator/language/en-GB/plg_task_checkfiles.ini (hunk 4 above)
Comment (bembelimen) [reply administrator/language/en-GB/plg_task_checkfiles.ini]: Thx

File: administrator/language/en-GB/plg_task_checkfiles.ini (hunk 2 above)
Comment (bembelimen) [reply administrator/language/en-GB/plg_task_checkfiles.ini]: Thx

File: plugins/task/checkfiles/checkfiles.php [hunk 5]
```diff
@@ -130,8 +132,14 @@ protected function checkImages(ExecuteTaskEvent $event): int
 				$this->logTask('PLG_TASK_CHECK_FILES_LOG_IMAGE_SAVE_FAIL', 'error');
 			}
 
-			// We do at most a single resize per execution
-			break;
+			--$numImages;
+
+			// We do a limited number of resize per execution
+
```
Comment (bembelimen) [inline plugins/task/checkfiles/checkfiles.php]:
```suggestion
```

File: plugins/task/checkfiles/forms/image_size.xml [hunk 6]
```diff
@@ -29,10 +30,21 @@
...
 				step="1"
 				filter="int"
 			/>
+			<field
+				name="numImages"
+				type="number"
+				label="PLG_TASK_CHECK_FILES_LABEL_DIMENSION_NUMIMAGES"
+				description="PLG_TASK_CHECK_FILES_LABEL_DIMENSION_NUMIMAGES_DESC"
+				required="true"
+				min="0"
```
Comment (Quy) [inline plugins/task/checkfiles/forms/image_size.xml]: Why `0`?

Comment (bembelimen) [inline plugins/task/checkfiles/forms/image_size.xml]:
```suggestion
				min="1"
```

File: administrator/language/en-GB/plg_task_checkfiles.ini [hunk 7]
```diff
@@ -5,12 +5,14 @@
 
 PLG_TASK_CHECK_FILES="Task - Check Files"
 PLG_TASK_CHECK_FILES_LABEL_DIMENSION_LIMIT="Limit (px)"
+PLG_TASK_CHECK_FILES_LABEL_DIMENSION_NUMIMAGES="Number of Images"
```
Comment (Quy) [inline administrator/language/en-GB/plg_task_checkfiles.ini]:
```suggestion
PLG_TASK_CHECK_FILES_LABEL_DIMENSION_NUMIMAGES="Maximum Images"
```
Probably change the key to not include `DIMENSION`.

File: plugins/task/checkfiles/checkfiles.php [hunk 8]
```diff
@@ -130,8 +132,13 @@ protected function checkImages(ExecuteTaskEvent $event): int
 				$this->logTask('PLG_TASK_CHECK_FILES_LOG_IMAGE_SAVE_FAIL', 'error');
 			}
 
-			// We do at most a single resize per execution
-			break;
+			--$numImages;
+
+			// We do a limited number of resize per execution
+			if ($numImages <= 0)
```
Comment (Quy) [inline plugins/task/checkfiles/checkfiles.php]:
```suggestion
			if ($numImages == 0)
```
Will it ever be less than 0?

Comment (bembelimen) [reply plugins/task/checkfiles/checkfiles.php]: No, you're right.

File: administrator/language/en-GB/plg_task_checkfiles.ini (hunk 7 above)
Comment (bembelimen) [reply administrator/language/en-GB/plg_task_checkfiles.ini]: Thx, improved the strings.

**Comment by chmst** [comment]
Now I get this error:
grafik

Folder Sample Data. Fresh installation of the full package.

**Comment by chmst** [comment]
Maybe only my local environment: ERROR	Routine threw exception: The GD extension for PHP is not available.

**Comment by drmenzelit** [comment]
I have tested this item :white_check_mark: successfully on 99bbefde68ca1c6555ed341a75093c627c9ed1b4<br /><br />Tested with 1 image, it works. Changed the settings to maximum images = 3, uploaded 4 big images, 3 where changed successfully on the next run. The 4th image was changed too in the next run.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36653">issues.joomla.org/tracker/joomla-cms/36653</a>.</sub>

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on 99bbefde68ca1c6555ed341a75093c627c9ed1b4<br /><br />Re-tested on web server and it works. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36653">issues.joomla.org/tracker/joomla-cms/36653</a>.</sub>

**Comment by drmenzelit** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36653">issues.joomla.org/tracker/joomla-cms/36653</a>.</sub>

**Comment by bembelimen** [comment]
Thanks all for the feedback and the improvements.

**Pull request merged**